flow/aggregates.py

```diff
@@ -0,0 +1,266 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Iterable
+from hashlib import md5
+from hashlib import sha1
+from itertools import groupby
+from itertools import zip_longest
+
+
+class Aggregate:
+    """Decorator for operation functions that are to be aggregated.
+    By default, if the aggregator parameter is not passed,
+    an aggregate of all jobs will be created.
+
+    .. code-block:: python
+
+        example_aggregate = Aggregate()
+        @example_aggregate
+        @FlowProject.operation
+        def foo(*jobs):
+            print(len(jobs))
+
+    :param aggregator:
+        Information describing how to aggregate jobs. Is a callable that
+        takes in a list of jobs and can return or yield subsets of jobs as
+        an iterable. The default behavior is creating a single aggregate
+        of all jobs
+    :type aggregator:
+        callable
+    :param sort_by:
+        Before aggregating, sort the jobs by a given statepoint parameter.
+        The default behavior is no sorting.
+    :type sort_by:
+        str or NoneType
+    :param reverse_order:
+        States if the jobs are to be sorted in reverse order.
+        The default value is False.
+    :type reverse_order:
+        bool
+    :param select:
+        Condition for filtering individual jobs. This is passed as the
+        callable argument to `filter`.
+        The default behavior is no filtering.
+    :type select:
+        callable or NoneType
+    """
+
+    def __init__(self, aggregator=None, sort_by=None, reverse_order=False, select=None):
+        if aggregator is None:
+            def aggregator(jobs):
+                return [jobs]
+
+        if not callable(aggregator):
+            raise TypeError(f"Expected callable for aggregator, got {type(aggregator)}")
+
+        if sort_by is not None and not isinstance(sort_by, str):
+            raise TypeError(f"Expected string sort_by parameter, got {type(sort_by)}")
+
+        if select is not None and not callable(select):
+            raise TypeError(f"Expected callable for select, got {type(select)}")
```

Comment:
All of this error checking feels rather un-Pythonic to me. However, I can see a good reason for this here since none of these attributes will actually get used until downstream in the aggregation pipeline, at which point the errors would be much harder for users to interpret. @bdice you're usually the style police for EAFP (I'm of the opinion that both LBYL and EAFP have their place in Python, so I'm less opposed to this on principle), what do you think?

Reply:
I am fine because this makes errors easier for the user to manage.